processor/src/main/java/org/mapstruct/ap/internal/model/BeanMappingMethod.java

```diff
@@ -844,20 +794,22 @@ private void applyParameterNameBasedMapping() {
             }
         }
 
-        private MappingOptions extractAdditionalOptions(String targetProperty, boolean restrictToDefinedMappings) {
-            MappingOptions additionalOptions = null;
+        private MappingReferences extractMappingReferences(String targetProperty, boolean restrictToDefinedMappings) {
             if ( unprocessedDefinedTargets.containsKey( targetProperty ) ) {
-                Set<Mapping> mappings = unprocessedDefinedTargets.get( targetProperty );
-                additionalOptions = MappingOptions.forMappingsOnly( mappings, restrictToDefinedMappings );
+                Set<MappingReference> mappings = unprocessedDefinedTargets.get( targetProperty );
+                return new MappingReferences( mappings, restrictToDefinedMappings );
             }
-            return additionalOptions;
+            return null;
         }
 
         private Accessor getTargetPropertyReadAccessor(String propertyName) {
             return method.getResultType().getPropertyReadAccessors().get( propertyName );
         }
 
         private ReportingPolicyPrism getUnmappedTargetPolicy() {
+            if ( mappingReferences.isForForgedMethods() ) {
```

Comment:
Apparently `isRestrictToDefinedMethods` is not the same as `isForForgedMethods`..

This part was copied form the `MappingOptions` (effectively setting the reporting policy to IGNORE via an empty `@BeanMapping` and the solution did not work when I did not introduce `isForForgedMethods`. 

I would expect `isForForgedMethods` and `isRestrictToDefinedMethods` to reflect the same state from a usage perspective: when restricting to defined methods, we should not report on remaining properties that could have been handled based on their name.

In short: I do not understand this piece of logic. @filiphr : can you shed some light on this?

Reply:
The reason why you need to do this is because now `method.getMappingOptions()` would return the mapping options on from the `basedOn` method.

Reply:
> The reason why you need to do this is because now `method.getMappingOptions()` would return the mapping options on from the `basedOn` method.

I understand that part.. But why (on a more fundamental level) is this different than isRestrictedToDefinedMethods.? Logically, you would not do error handling (override IGNORE in `method.getMappingOptions()`)  you would expect to do this when you `restrictToDefinedMethods` because you don't want to report errors on 'missed' name based mappings.

Reply:
For the `ForgedMethod` we are crafting new `MappingOptions`. It doesn't make sense to reuse the `MappingOptions` from the `basedOn` method. I think that `restrictToDefinedMethods` was added to support the complex target nested mappings. This is not exposed over annotations.

Reply:
The `MappingOptions` filter information in hierarchy. I can agree that `@BeanMapping` is not relevant, because generated methods can potentially be shared between defined mapping methods. But stuff in `@Mapper`, `@MappingConfig` is relevant..

You could even make a case for trickling down `@BeanMapping` stuff to generated methods. Think about qualifiers for before/after mappings. It is a bit tricky, because that would mean that a forged method cannot be shared if the `@MeanMapping` is different on some level.

Reply:
`@MeanMapping` == of course `@BeanMapping`.. but it can be mean :smile:

Reply:
We do share the stuff from `@Mapper`, `@MappingConfig`, since we get those from the `MapperType`. For trickling down `@BeanMapping`, yes you can make a case (and I think that we have an issue to do that for `ForgedMethods`. However, you can't trickle down everything. `BeanMapping#resultType` for example. 

You can even argue that if someone wants to define some things specific for some forged methods that this should be either done via `@Mapper` or maybe they should define this method in their mapper.

Reply:
> However, you can't trickle down everything. `BeanMapping#resultType` for example.

Correct. And to be honest, I don't think we should do this.. (hand down `@BeanMapping`).. If there's an issue on that, we should answer it with this argument that forged methods can be shared at some level, and its hard to see which method originally created it.